pkg/sigv4/sigv4.go

```diff
@@ -0,0 +1,200 @@
+package sigv4
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+)
+
+const signingAlgo = "AWS4-HMAC-SHA256"
+const awsServiceName = "aps"
+
+type Signer interface {
+	Sign(req *http.Request) error
+}
+
+type DefaultSigner struct {
+	iSO8601Date      string
+	canonicalHeaders string
+	signedHeaders    string
+	credentialScope  string
+	config           *Config
+	payloadHash      string
+}
+
+func NewDefaultSigner(config *Config) Signer {
+	return &DefaultSigner{
+		config: config,
+	}
+}
+
+func (d *DefaultSigner) Sign(req *http.Request) error {
+	now := time.Now().UTC()
+	d.iSO8601Date = now.Format("20060102T150405Z")
+	d.credentialScope = fmt.Sprintf(
+		"%s/%s/%s/aws4_request",
+		now.UTC().Format("20060102"),
+		d.config.Region,
+		awsServiceName,
+	)
+
+	payloadHash, err := d.getPayloadHash(req)
+	if err != nil {
+		return err
+	}
+
+	d.payloadHash = payloadHash
+	d.addRequiredHeaders(req)
+	d.canonicalHeaders, d.signedHeaders = d.getCanonicalAndSignedHeaders(req)
```

Comment:
Is there a big reason to keep the state (`canonicalHeaders` and `signedHeaders` and request related values) on the singer level instead of passing them downstream?

Reply:
They were used in multiple places and I didn't want to pass them individually as arguments. 

I changed it as per your suggestion.